### PR TITLE
simplify and dry up job execution logic and queing

### PR DESCRIPTION
**Pull request by grosser**
this started as a simple refactor ... but I got drawn deeper and deeper into the rabbit hole ...

 - clearer names
 - JobQueue is responsible for queuing, others do not know about internals
 - fix memory leak in JobQueue
 - reuse 'waiting for restart' logic
 - fix that executing but inactive jobs were not shown
 - fix new jobs during restarts being stored in registry and never getting removed
 - full test coverage for JobQueue

@sandlerr @jonmoter @anthonywoo

### CODE REVIEW

File: app/models/deploy.rb [hunk 1]
```diff
@@ -27,9 +27,10 @@ def cache_key
     [super, commit]
   end
 
-  def job_execution_key
-    # if stage can run in parallel - set key to a unique id so it can immediately execute
-    stage.run_in_parallel ? "deploy-#{id}" : "stage-#{stage.id}"
+  # use a unique queue for stages that can execute in parallel so they start immediately
+  # job_id is what JobExecution also internally uses as queue for 'normal' jobs
+  def job_execution_queue
```
Comment (sandlerr) [inline app/models/deploy.rb]: `job_execution_queue_name` since it doesn't return a `JobQueue` instance

File: app/models/job_queue.rb [hunk 2]
```diff
@@ -1,55 +1,48 @@
...
-  def active?(key, id)
-    @active[key].try(:id) == id
+  def active?(id)
+    active.detect { |je| je.id == id }
   end
 
-  def queued?(key, id)
-    @queue.dup[key].any? { |je| je.id == id }
+  def queued?(id)
+    @queue.values.detect { |jes| jes.detect { |je| return je if je.id == id } }
```
Comment (sandlerr) [inline app/models/job_queue.rb]: wasn't `any?` better here and in `active?` I'd expect this to return boolean

**Comment by grosser** [comment]
Hehe yeah needed the return value for find-by-id :)

**Comment by grosser** [comment]
green locally and green in docker ... damn it :/

File: app/models/job_queue.rb (hunk 2 above)
Comment (grosser) [inline app/models/job_queue.rb]: It's either having a `queued(id)` and `queued?(id)` method or this non-boolean `?` method ... you can pick ... I don't care much :D

**Comment by grosser** [comment]
found it ... multilock was left open and test order changed ... good fun ...all green now :)

File: app/models/job_queue.rb (hunk 2 above)
Comment (sandlerr) [inline app/models/job_queue.rb]: I'd go for having both methods to be more conventional, not a big deal.

**Comment by grosser** [comment]
hope these lock / job fixes make our build more stable too ...

**Pull request merged**